// _@jscpd_core@3.3.0-rc.8@@jscpd/core/dist/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getDefaultOptions() {
    return {
        executionId: new Date().toISOString(),
        path: [process.cwd()],
        minLines: 5,
        maxLines: 500,
        maxSize: '30kb',
        minTokens: 50,
        output: './report',
        reporters: ['console'],
        ignore: [],
        mode: 'mild',
        threshold: undefined,
        formatsExts: {},
        debug: false,
        silent: false,
        blame: false,
        cache: true,
        absolute: false,
        noSymlinks: false,
        skipLocal: false,
        ignoreCase: false,
        gitignore: false,
        reportersOptions: {},
    };
}
exports.getDefaultOptions = getDefaultOptions;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getOption(name, options) {
    const defaultOptions = getDefaultOptions();
    return options ? options[name] || defaultOptions[name] : defaultOptions[name];
}
exports.getOption = getOption;
