// _jscpd@3.3.0-rc.9@jscpd/dist/init/reporters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const finder_1 = require("@jscpd/finder");
const safe_1 = require("colors/safe");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const reporters = {
    xml: finder_1.XmlReporter,
    json: finder_1.JsonReporter,
    consoleFull: finder_1.ConsoleFullReporter,
    html: finder_1.HtmlReporter,
    console: finder_1.ConsoleReporter,
    silent: finder_1.SilentReporter,
    threshold: finder_1.ThresholdReporter,
};
function registerReporters(options, detector) {
    options.reporters.forEach((reporter) => {
        if (reporter in reporters) {
            detector.registerReporter(new reporters[reporter](options));
        }
        else {
            try {
                const reporterClass = require(`jscpd-${reporter}-reporter`).default;
                detector.registerReporter(new reporterClass(options));
            }
            catch (e) {
                try {
                    const reporterClass = require(`@jscpd/${reporter}-reporter`).default;
                    detector.registerReporter(new reporterClass(options));
                }
                catch (e) {
                    console.log(safe_1.yellow(`warning: ${reporter} not installed (install packages named @jscpd/${reporter}-reporter or jscpd-${reporter}-reporter)`));
                    console.log(safe_1.grey(e.message));
                }
            }
        }
    });
}
exports.registerReporters = registerReporters;
